src/db/index.rs

```diff
@@ -187,80 +153,209 @@ fn get_index_keys(
     Ok(vec![match target {
         // TODO: Support array of strings here.
         // TODO: when we support arrays, add test to ensure that strings with null are skipped.
-        Value::String(v) => {
-            if v.contains('\0') {
-                return Err(StringContainsNull(v.clone()));
-            }
-            index_key(IndexValue::Str(&v), key)?
-        }
+        Value::String(v) => encode_index_key(&IndexKey {
+            secondary: v,
+            primary: key,
+        })?,
         _ => return Err(UnsupportedTargetType),
     }])
 }
 
-// Returns the index key (index prolly map key) for an indexed value.
-pub fn index_key(secondary: IndexValue, primary: &[u8]) -> Result<Vec<u8>, GetIndexKeysError> {
+static KEY_VERSION_0: &[u8] = &[0];
+static KEY_SEPARATOR: &[u8] = &[0];
+
+// An index key is encoded to vec of bytes in the following order:
+//   - key version byte(s), followed by
+//   - the UTF8 encoded secondary key value, followed by
+//   - the key separator, a null byte, followed by
+//   - the primary key bytes
+//
+// The null separator byte ensures that if a secondary key A is longer than B then
+// A always sorts after B. Appending the primary key ensures index keys with
+// identical secondary keys sort in primary key order. Secondary keys must not
+// contain a null byte.
+pub fn encode_index_key(index_key: &IndexKey) -> Result<Vec<u8>, GetIndexKeysError> {
     use GetIndexKeysError::*;
 
-    let k = IndexKeyType::V0(IndexKey { secondary, primary });
-    Ok(bytekey::serialize(&k).map_err(|e| SerializeIndexEntryError(e.to_string()))?)
+    let IndexKey { secondary, primary } = *index_key;
+
+    if secondary.contains('\0') {
+        return Err(StringContainsNull(secondary.to_string()));
+    }
+    let mut v = Vec::new();
+    v.reserve(KEY_VERSION_0.len() + secondary.len() + KEY_SEPARATOR.len() + primary.len());
+    v.extend_from_slice(KEY_VERSION_0);
+    v.extend_from_slice(secondary.as_bytes());
+    v.extend_from_slice(KEY_SEPARATOR);
+    v.extend_from_slice(primary);
+    Ok(v)
 }
 
 // Returns bytes that can be used to scan for the given secondary index value. These bytes
-// are different from the full index key bytes in that we do *not* want the trailing null
-// byte that bytekey appends to the secondary index value ("key\0" is not a prefix of "keyfoo").
-pub fn scan_key(secondary: &str) -> Result<Vec<u8>, GetIndexKeysError> {
-    let mut k = index_key(IndexValue::Str(secondary), &[])?;
-    k.truncate(k.len() - 1);
+// are different from the full index key bytes in that we do *not* want the null
+// byte separating secondary and primary keys ("key\0" is not a prefix of "keyfoo").
+// Exclusive scans are implemented by appending a 0x01 byte to the end of the scan key,
+// making it start atthe next higher value.
```

Comment:
typo

Reply:
due to several comment overhauls this paragraph got nuked